core/meterer/onchain_state.go

```diff
@@ -106,63 +113,84 @@ func (pcs *OnchainPaymentState) RefreshOnchainPaymentState(ctx context.Context,
 		return err
 	}
 	// These parameters should be rarely updated, but we refresh them anyway
-	pcs.PaymentVaultParams = paymentVaultParams
+	pcs.PaymentVaultParams.Store(paymentVaultParams)
+
+	pcs.ReservationsLock.Lock()
+	accountIDs := make([]gethcommon.Address, 0, len(pcs.ActiveReservations))
+	for accountID := range pcs.ActiveReservations {
+		accountIDs = append(accountIDs, accountID)
+	}
+
+	activeReservations, err := tx.GetActiveReservations(ctx, accountIDs)
+	if err != nil {
+		return err
+	}
+	pcs.ActiveReservations = activeReservations
+	pcs.ReservationsLock.Unlock()
+
+	pcs.OnDemandLocks.Lock()
+	accountIDs = make([]gethcommon.Address, 0, len(pcs.OnDemandPayments))
+	for accountID := range pcs.OnDemandPayments {
+		accountIDs = append(accountIDs, accountID)
+	}
+
+	onDemandPayments, err := tx.GetOnDemandPayments(ctx, accountIDs)
+	if err != nil {
+		return err
+	}
+	pcs.OnDemandPayments = onDemandPayments
+	pcs.OnDemandLocks.Unlock()
+
 	return nil
 }
 
 // GetActiveReservationByAccount returns a pointer to the active reservation for the given account ID; no writes will be made to the reservation
-func (pcs *OnchainPaymentState) GetActiveReservationByAccount(ctx context.Context, accountID string) (core.ActiveReservation, error) {
-	if reservation, ok := pcs.ActiveReservations[accountID]; ok {
+func (pcs *OnchainPaymentState) GetActiveReservationByAccount(ctx context.Context, accountID gethcommon.Address) (*core.ActiveReservation, error) {
+	if reservation, ok := (pcs.ActiveReservations)[accountID]; ok {
```

Comment:
Don't we need rlock here? I think it can crash if a map is read at the same time it's being written

Reply:
oof, I added RLock and defer RUnlock

Reply:
If you `defer pcs.ReservationsLock.RUnlock()` it will hold the rlock until the end of the execution. Will it cause conflict with write lock in L160?